File: apps/ios/Voiid/Voiid/Networking/GoogleDriveBackupService.swift
//
//  GoogleDriveBackupService.swift
//  Voiid
//
//  Google Drive destination for the encrypted backup blob. The blob stored here is
//  the exact `encryptBackup(masterSecret, plaintext)` ciphertext — Google only ever
//  sees the opaque blob; the master secret and plaintext never leave the device.
//
//  The file lives in the Drive **appDataFolder** — a hidden, app-private folder that
//  the `https://www.googleapis.com/auth/drive.appdata` scope grants. That scope is
//  app-private: it does NOT see or touch the user's other Drive files. Fixed name
//  `voiid-backup.enc`; we PATCH the existing file id in place (no duplicates).
//
//  --- Dependency / SDK path taken ---------------------------------------------------
//  The Drive v3 REST calls below are implemented directly on URLSession (the same
//  raw-REST approach BackupService/MediaService already use for R2), so this file is
//  real, compiling Drive logic with NO third-party SDK required. The one thing REST
//  cannot do by itself is acquire the OAuth access token — that's what the GoogleSignIn
//  SDK provides. Rather than hand-wire GoogleSignIn-iOS's multi-package graph (AppAuth,
//  GTMAppAuth, GTMSessionFetcher) into the project here, token acquisition is abstracted
//  behind `GoogleDriveTokenProvider`. The default provider is a stub that reports
//  "not signed in", so the feature compiles and is safely disabled until GoogleSignIn
//  is wired. TODO(GoogleSignIn): add the SPM package `https://github.com/google/GoogleSignIn-iOS`,
//  request the `drive.appdata` scope, and set `GoogleDriveBackupService.shared.tokenProvider`
//  to an adapter whose `accessToken()` returns `GIDSignIn ... .accessToken.tokenString`
//  (refreshing as needed). The reversed-client-id URL scheme is already in Info.plist.
//

import Foundation

// MARK: - OAuth token provider (the seam GoogleSignIn fills)

/// Supplies a valid Drive access token scoped to `drive.appdata`. GoogleSignIn implements
/// this once it's wired; until then the stub below reports "not signed in".
protocol GoogleDriveTokenProvider: Sendable {
    /// Whether a Google account is currently authorized for Drive backup.
    var isSignedIn: Bool { get }
    /// A fresh, valid OAuth access token, refreshing if necessary. Throws if not signed in.
    func accessToken() async throws -> String
}

/// Default provider used until GoogleSignIn is integrated: always "signed out". This keeps
/// the whole Drive feature compiling and visibly disabled (no client secret, no network).
struct StubGoogleDriveTokenProvider: GoogleDriveTokenProvider {
    var isSignedIn: Bool { false }
    func accessToken() async throws -> String { throw GoogleDriveError.signInRequired }
}

// MARK: - Errors

enum GoogleDriveError: LocalizedError {
    case signInRequired
    case noBackup
    case http(status: Int, message: String)
    case badResponse

    var errorDescription: String? {
        switch self {
        case .signInRequired:            return "Sign in with Google to use Google Drive backup."
        case .noBackup:                  return "No Google Drive backup was found."
        case .http(let status, let msg): return "Google Drive request failed (\(status)). \(msg)"
        case .badResponse:               return "Unexpected response from Google Drive."
        }
    }
}

// MARK: - Service

/// Drive v3 REST client for the encrypted backup blob, targeting `appDataFolder`.
///
/// `@unchecked Sendable`: `tokenProvider` is only swapped once at startup (when GoogleSignIn
/// is wired); all network work is stateless URLSession calls.
final class GoogleDriveBackupService: BackupDestinationService, @unchecked Sendable {
    static let shared = GoogleDriveBackupService()
    private init() {}

    /// Injected by GoogleSignIn wiring; defaults to the "signed out" stub.
    var tokenProvider: GoogleDriveTokenProvider = StubGoogleDriveTokenProvider()

    /// Whether Drive backup can run right now (a Google account is authorized).
    var isSignedIn: Bool { tokenProvider.isSignedIn }

    /// The app-private Drive scope. Least-privilege: sees only our own appDataFolder, never
    /// the user's other files. Referenced by the (future) GoogleSignIn authorization request.
    static let scope = "https://www.googleapis.com/auth/drive.appdata"

    /// Fixed backup filename inside appDataFolder.
    private static let filename = "voiid-backup.enc"

    private static let apiBase = "https://www.googleapis.com/drive/v3"
    private static let uploadBase = "https://www.googleapis.com/upload/drive/v3"

    // MARK: BackupDestinationService

    /// Create-or-update the backup file. If it already exists we PATCH its bytes in place
    /// (no duplicates); otherwise we multipart-create it in `appDataFolder`. Always the
    /// encrypted ciphertext blob.
    func uploadBackup(_ blob: Data) async throws {
        let token = try await tokenProvider.accessToken()
        if let existing = try await findBackupFile(token: token) {
            try await updateFile(id: existing.id, blob: blob, token: token)
        } else {
            try await createFile(blob: blob, token: token)
        }
    }

    func fetchSnapshot() async throws -> BackupSnapshot? {
        // Not signed in → destination simply unavailable (nil), not an error.
        guard tokenProvider.isSignedIn else { return nil }
        let token = try await tokenProvider.accessToken()
        guard let file = try await findBackupFile(token: token) else { return nil }
        let size = Int(file.size ?? "0") ?? 0
        let modified = file.modifiedTime.flatMap { ISO8601DateFormatter().date(from: $0) }
        return BackupSnapshot(sizeBytes: size, modified: modified)
    }

    func downloadBackup() async throws -> Data {
        let token = try await tokenProvider.accessToken()
        guard let file = try await findBackupFile(token: token) else { throw GoogleDriveError.noBackup }
        var req = URLRequest(url: url("\(Self.apiBase)/files/\(file.id)", query: [("alt", "media")]))
        req.httpMethod = "GET"
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        let (data, resp) = try await send(req)
        try Self.ensureOK(resp, data)
        return data
    }

    /// Delete the Drive backup (used by "disable / delete Drive backup"). No-op if none.
    func deleteBackup() async throws {
        let token = try await tokenProvider.accessToken()
        guard let file = try await findBackupFile(token: token) else { return }
        var req = URLRequest(url: url("\(Self.apiBase)/files/\(file.id)"))
        req.httpMethod = "DELETE"
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        let (data, resp) = try await send(req)
        try Self.ensureOK(resp, data)
    }

    // MARK: REST helpers

    private struct DriveFile: Decodable { let id: String; let name: String?; let size: String?; let modifiedTime: String? }
    private struct DriveFileList: Decodable { let files: [DriveFile] }

    /// List appDataFolder for our fixed-name file. Returns nil if absent.
    private func findBackupFile(token: String) async throws -> DriveFile? {
        var req = URLRequest(url: url("\(Self.apiBase)/files", query: [
            ("spaces", "appDataFolder"),
            ("q", "name = '\(Self.filename)'"),
            ("fields", "files(id,name,size,modifiedTime)"),
            ("pageSize", "1"),
        ]))
        req.httpMethod = "GET"
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        let (data, resp) = try await send(req)
        try Self.ensureOK(resp, data)
        let list = try JSONDecoder().decode(DriveFileList.self, from: data)
        return list.files.first
    }

    /// Multipart create in appDataFolder (metadata part + media part).
    private func createFile(blob: Data, token: String) async throws {
        let boundary = "voiid-\(UUID().uuidString)"
        let metadata = #"{"name":"\#(Self.filename)","parents":["appDataFolder"]}"#

        var body = Data()
        body.append("--\(boundary)\r\n".utf8Data)
        body.append("Content-Type: application/json; charset=UTF-8\r\n\r\n".utf8Data)
        body.append(metadata.utf8Data)
        body.append("\r\n--\(boundary)\r\n".utf8Data)
        body.append("Content-Type: application/octet-stream\r\n\r\n".utf8Data)
        body.append(blob)
        body.append("\r\n--\(boundary)--\r\n".utf8Data)

        var req = URLRequest(url: url("\(Self.uploadBase)/files", query: [("uploadType", "multipart")]))
        req.httpMethod = "POST"
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        req.setValue("multipart/related; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        req.httpBody = body
        let (data, resp) = try await send(req)
        try Self.ensureOK(resp, data)
    }

    /// PATCH the existing file's bytes in place (media upload; metadata/name unchanged).
    private func updateFile(id: String, blob: Data, token: String) async throws {
        var req = URLRequest(url: url("\(Self.uploadBase)/files/\(id)", query: [("uploadType", "media")]))
        req.httpMethod = "PATCH"
        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        req.setValue("application/octet-stream", forHTTPHeaderField: "Content-Type")
        req.httpBody = blob
        let (data, resp) = try await send(req)
        try Self.ensureOK(resp, data)
    }

    private func url(_ base: String, query: [(String, String)] = []) -> URL {
        var comps = URLComponents(string: base)!
        if !query.isEmpty { comps.queryItems = query.map { URLQueryItem(name: $0.0, value: $0.1) } }
        return comps.url!
    }

    private func send(_ req: URLRequest) async throws -> (Data, URLResponse) {
        do { return try await URLSession.shared.data(for: req) }
        catch { throw GoogleDriveError.http(status: 0, message: error.localizedDescription) }
    }

    private static func ensureOK(_ resp: URLResponse, _ data: Data) throws {
        let status = (resp as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(status) else {
            let msg = String(data: data, encoding: .utf8) ?? ""
            throw GoogleDriveError.http(status: status, message: String(msg.prefix(200)))
        }
    }
}

private extension String {
    var utf8Data: Data { Data(utf8) }
}

File: apps/android/app/src/main/java/com/voiid/app/net/GoogleDriveBackupService.kt
package com.voiid.app.net

import android.content.Context
import android.content.Intent
import com.google.android.gms.auth.GoogleAuthUtil
import com.google.android.gms.auth.UserRecoverableAuthException
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInOptions
import com.google.android.gms.common.api.Scope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.ByteArrayOutputStream
import java.net.URLEncoder
import java.util.concurrent.TimeUnit

/**
 * SECOND, user-owned destination for the SAME encrypted backup blob that
 * [BackupService] pushes to our server. The bytes written here are ALWAYS the
 * `encryptBackup(masterSecret, plaintext)` ciphertext produced in [BackupManager];
 * the master secret and plaintext never leave the device — Google only ever sees
 * opaque ciphertext.
 *
 * Storage is the Drive **`appDataFolder`** — a hidden, app-private folder the user can't
 * browse and other apps can't read. That's why we request the least-privilege
 * `drive.appdata` scope (NOT full Drive): Google can't see the user's other files, and
 * this app can only touch what it created.
 *
 * Transport is Drive v3 REST over the existing OkHttp (mirrors how [BackupService] hits
 * presigned R2 directly) — no heavy Drive client library. The OAuth access token comes
 * from [DriveTokenProvider]; the default implementation uses Google Sign-In +
 * [GoogleAuthUtil]. Sign-in requires the OAuth Android client id / SHA registered in the
 * Google console (docs/GOOGLE_INTEGRATION_TODO.md §2a) — until that console step is done
 * sign-in fails at runtime, but the code compiles and the REST logic is real.
 *
 * All calls are blocking network/crypto work wrapped in [Dispatchers.IO]; call from a
 * coroutine. Failures throw [ApiError.Http] (HTTP status) or propagate the Google auth
 * exception; [DriveAuthRecoverable] carries a consent [Intent] the UI can launch.
 */
class GoogleDriveBackupService(
    context: Context,
    private val tokenProvider: DriveTokenProvider =
        GoogleSignInTokenProvider(context.applicationContext),
) {
    private val appContext = context.applicationContext

    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .writeTimeout(120, TimeUnit.SECONDS)
        .readTimeout(120, TimeUnit.SECONDS)
        .build()

    /** Metadata for our backup file in the Drive appDataFolder (mirrors [BackupService.BackupMeta]). */
    data class DriveBackupMeta(
        val fileId: String,
        val sizeBytes: Long,
        val modifiedTime: String?,
    )

    @Serializable private data class DriveFile(
        val id: String,
        val name: String? = null,
        val size: String? = null,        // Drive v3 returns size as a decimal string
        val modifiedTime: String? = null,
    )
    @Serializable private data class DriveFileList(val files: List<DriveFile> = emptyList())

    // MARK: - Sign-in surface (used by the UI to launch the consent flow)

    /** Google Sign-In client scoped to `drive.appdata`. The UI launches [GoogleSignInClient.getSignInIntent]. */
    fun signInClient(): GoogleSignInClient {
        val opts = GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
            .requestEmail()
            .requestScopes(Scope(DRIVE_APPDATA_SCOPE))
            .build()
        return GoogleSignIn.getClient(appContext, opts)
    }

    /** True if there's a signed-in Google account that granted the `drive.appdata` scope. */
    fun isSignedIn(): Boolean {
        val acct = GoogleSignIn.getLastSignedInAccount(appContext) ?: return false
        return GoogleSignIn.hasPermissions(acct, Scope(DRIVE_APPDATA_SCOPE))
    }

    /** Sign out of Google on this device (revokes local token; the Drive blob itself is untouched). */
    fun signOut() {
        runCatching { signInClient().signOut() }
    }

    // MARK: - Blob transport (same shapes as BackupService)

    /** Upload the encrypted backup blob to appDataFolder — updating the existing file id if present. */
    suspend fun uploadBackup(blob: ByteArray) = withContext(Dispatchers.IO) {
        val token = requireToken()
        val existing = findBackupFile(token)
        if (existing != null) updateMedia(token, existing.id, blob)
        else createMultipart(token, blob)
    }

    /** Metadata for our backup file, or null if not signed in / no backup yet. */
    suspend fun fetchBackupMeta(): DriveBackupMeta? = withContext(Dispatchers.IO) {
        val token = tokenOrNull() ?: return@withContext null
        val f = findBackupFile(token) ?: return@withContext null
        DriveBackupMeta(f.id, f.size?.toLongOrNull() ?: 0L, f.modifiedTime)
    }

    /** Download the encrypted backup blob (`GET files/{id}?alt=media`). */
    suspend fun downloadBackup(): ByteArray = withContext(Dispatchers.IO) {
        val token = requireToken()
        val f = findBackupFile(token) ?: throw ApiError.Http(404, "No Google Drive backup found.")
        val req = Request.Builder()
            .url("https://www.googleapis.com/drive/v3/files/${f.id}?alt=media")
            .header("Authorization", "Bearer $token")
            .get().build()
        execOrThrow(req, "Drive download failed")
    }

    // MARK: - Drive v3 REST internals

    private fun findBackupFile(token: String): DriveFile? {
        val q = URLEncoder.encode("name = '$BACKUP_FILE_NAME'", "UTF-8")
        val fields = URLEncoder.encode("files(id,name,size,modifiedTime)", "UTF-8")
        val req = Request.Builder()
            .url("https://www.googleapis.com/drive/v3/files?spaces=appDataFolder&q=$q&fields=$fields")
            .header("Authorization", "Bearer $token")
            .get().build()
        val bytes = execOrThrow(req, "Drive list failed")
        return ApiClient.json.decodeFromString(DriveFileList.serializer(), String(bytes)).files.firstOrNull()
    }

    /** New file: multipart/related (metadata part pins parents=appDataFolder + name, then the media part). */
    private fun createMultipart(token: String, blob: ByteArray) {
        val boundary = "voiid_" + System.currentTimeMillis()
        val meta = """{"name":"$BACKUP_FILE_NAME","parents":["appDataFolder"]}"""
        val body = ByteArrayOutputStream().apply {
            fun w(s: String) = write(s.toByteArray(Charsets.UTF_8))
            w("--$boundary\r\n")
            w("Content-Type: application/json; charset=UTF-8\r\n\r\n")
            w(meta); w("\r\n")
            w("--$boundary\r\n")
            w("Content-Type: application/octet-stream\r\n\r\n")
            write(blob); w("\r\n")
            w("--$boundary--")
        }.toByteArray()
        val req = Request.Builder()
            .url("https://www.googleapis.com/upload/drive/v3/files?uploadType=multipart&fields=id")
            .header("Authorization", "Bearer $token")
            .post(body.toRequestBody("multipart/related; boundary=$boundary".toMediaType()))
            .build()
        execOrThrow(req, "Drive upload failed")
    }

    /** Existing file: replace its bytes only (media upload; parents/name unchanged). */
    private fun updateMedia(token: String, fileId: String, blob: ByteArray) {
        val req = Request.Builder()
            .url("https://www.googleapis.com/upload/drive/v3/files/$fileId?uploadType=media&fields=id")
            .header("Authorization", "Bearer $token")
            .patch(blob.toRequestBody("application/octet-stream".toMediaType()))
            .build()
        execOrThrow(req, "Drive update failed")
    }

    private fun execOrThrow(req: Request, ctx: String): ByteArray {
        client.newCall(req).execute().use {
            if (!it.isSuccessful) throw ApiError.Http(it.code, "$ctx (${it.code}).")
            return it.body?.bytes() ?: ByteArray(0)
        }
    }

    private fun requireToken(): String =
        tokenOrNull() ?: throw ApiError.Http(401, "Sign in to Google Drive first.")

    /** Blocking OAuth token (we're already on Dispatchers.IO); null if no signed-in account. */
    private fun tokenOrNull(): String? = tokenProvider.blockingToken()

    companion object {
        /** Least-privilege Drive scope: app-private folder only — Google can't see the user's other files. */
        const val DRIVE_APPDATA_SCOPE = "https://www.googleapis.com/auth/drive.appdata"
        private const val BACKUP_FILE_NAME = "voiid-backup.enc"
    }
}

/**
 * Seam for OAuth-token acquisition, so the Drive REST logic doesn't hard-depend on the
 * sign-in mechanism. The default [GoogleSignInTokenProvider] uses Google Sign-In +
 * [GoogleAuthUtil]; tests/alt-auth can swap it.
 */
interface DriveTokenProvider {
    /**
     * A fresh OAuth access token for the `drive.appdata` scope, or null if no Google
     * account is signed in on this device. Blocking — call on [Dispatchers.IO].
     * Throws [DriveAuthRecoverable] if the user must (re)consent.
     */
    fun blockingToken(): String?
}

/** Signals the UI to launch [recoveryIntent] to (re)obtain the user's Drive consent. */
class DriveAuthRecoverable(val recoveryIntent: Intent) :
    Exception("Google authorization required")

/** Default provider: reads the last signed-in Google account and mints a token via [GoogleAuthUtil]. */
class GoogleSignInTokenProvider(context: Context) : DriveTokenProvider {
    private val appContext = context.applicationContext

    override fun blockingToken(): String? {
        val account = GoogleSignIn.getLastSignedInAccount(appContext)?.account ?: return null
        return try {
            GoogleAuthUtil.getToken(
                appContext,
                account,
                "oauth2:${GoogleDriveBackupService.DRIVE_APPDATA_SCOPE}",
            )
        } catch (e: UserRecoverableAuthException) {
            val intent = e.intent ?: throw e
            throw DriveAuthRecoverable(intent)
        }
    }
}
